### ISSUE TITLE
Reason for slow queries?

### ISSUE THREAD

**Issue by sdegutis** [pejulian/romcal#236]
Getting today's feast, by limiting it to the current month, and then scanning for the current day, takes about a second. What's the reason for this performance? I'm running it in the browser fwiw, so it shouldn't be accessing files or other IO.

**Comment by emagnier** [comment]
Romcal must compute first the whole liturgical year, even if you need then data for only day. Because of different rules and criteria where liturgical days can move around according to each others and the civil calendar.

Indeed there is still place for code optimisation to reduce a bit the computing time. Also ou next major version will be async and will expose more utility functions to manipulate and retrieve liturgical dates, without necessarily computing everything (it's almost ready for a beta, I just can't find time to do the last miles). In the meantime you could already wrap romcal in an async function and cache the results, if that could help in your implementation.

Thanks for your feedback!

**Comment by sdegutis** [comment]
Why will the next major version be async? Is it because of the problem described in this issue?

Looking through the code, I didn't realize how extensively lodash is used. Normally this kind of lag is only expected when there's some kind of I/O which is what I looked for first, but constant creation and invocation of functions to this extent can add up too. Extra code due to bundling probably also plays a role.

I think just rewriting the code to use for-loops and be more O(1) with the data might make the code near instantaneous and get rid of the need for an async API (assuming that's the reason).

**Comment by tukusejssirs** [comment]
@sdegutis, I am not a romcal coder (@emagnier is), however, you can look at the latest dev code in the [`_F-TypeScript`](https://github.com/romcal/romcal/tree/_F-TypeScript) branch, or even more recent (unmerged) in #213. From what I know, the code in that branch is async.

**Comment by emagnier** [comment]
The next version will be async essentially for lazy load resources (calendar and localisation files). However  this feature is still at an early stage and will need more contributions to be fully effective.

One of the goals for this next release is the codebase converted to TypeScript. Several code refactor have been already done, but once released (at least a beta), the TypeScript language will be then a great help to do more refactor and code optimisation. Removing lodash to native JS features is also our plan.
I hope to release a first beta version within the next month (but I can't promise yet), then future contributions will be easier and more than welcome.

**Comment by sdegutis** [comment]
I might be interested in contributing, having experience with all these things. However, the codebase is confusing, and I'm having a hard time finding the source of truth of the data and algorithms.

**Comment by emagnier** [comment]
Yes actually the codebase is not in an ideal stage. The latest version is in this branch https://github.com/romcal/romcal/tree/_F-TypeScript, features and APIs could move a little further, and a big update and cleanup about content and localisation is also coming.

The blocker for a beta release (against the `dev` branch) is this big content update (currently being done in https://github.com/romcal/romcal/pull/213), and eventually a redesign of an experimental feature to extend calendars together (although this last feature could be ignored or removed for now, and completed in a next beta).

Please have a look to this branch. I also tried to update the documentation accordingly (if it can help).
Thank you for your offer of contribution 👍

**Comment by sdegutis** [comment]
Thanks for the info. I have a personal interest in this issue because I use this library on a website and in an app that our family uses. But I don't know anything about the algorithms needed to calculate such calendars as this, so it will take a while to dive into it and understand it before being able to help with it. Besides that, I am very busy and won't have spare time for this for a few weeks at least. So just letting you know that I'm interested in helping with this, but won't be available for a while.

**Comment by tukusejssirs** [comment]
@sdegutis, we all are short of time that could be offered to devolop romcal. :smiley:

Nevertheless, a kind of workaround is (as @emagnier somewhere already suggested) to cache the romcal output object. Its content only changes when (1) you change the input (the romcal options, such as locale or calendar or year); or (2) when you update romcal. On the webserver, I believe you can have this object in memory (in RAM or HDD) and query it instead of using romcal directly. I have no idea what kind of app you and your family use (OS, programming language), but I think the same algorithm could be used in there, too.

**Comment by sdegutis** [comment]
We use https://github.com/sdegutis/floating-clock currently. Though I plan to make a better version that allows embedding holy pictures (e.g. Sacred Heart and Immaculate Heart) and is built on Web Components.

The problem with caching it is that we need to change the input, i.e. the country ID when we toggle through the countries to see if there's an interesting feast day for the day. For example when St. John Henry Newman's feast day was on, the US and General calendars didn't mention it, but the England one did. So there's nothing to cache, as the input is what's changing. Thus it takes a good 1.5 seconds to load every time we click the arrow. We could also just load all the country codes all at once, but that freezes the whole click for 20 seconds.

**Comment by sdegutis** [comment]
Is there a general description of the algorithm this repo uses? It seems there's two main parts: the data files, which this repo mostly has; and the algorithm that uses those data files and calculates things. I kind of wonder if it might be simpler to start from the algorithm than from the code, to try to understand it and perhaps write a greenfield solution making use of the current data files. (Maybe that's what that TypeScript branch was already doing?)

**Comment by tukusejssirs** [comment]
> The problem with caching it is that we need to change the input, i.e. the country ID when we toggle through the countries to see if there's an interesting feast day for the day.

You could cache all the supported calendars (or just a subset of them), though it might have its size. :shrug:

> Is there a general description of the algorithm this repo uses?

@emagnier (and @pejulian) could answer this better. From what I know, not all portions of the code are fully described.

I personally don’t code romcal. I am (sort of) l10n head, calendar data (celebrations) proofer, ideas ‘sharer’ and issue commenter. :smiley:

Anyway, I think you should use (if you don’t use already) the code from the [`_F-TypeScript`](https://github.com/romcal/romcal/tree/_F-TypeScript) branch. Although it is a dev version, it contains many, many fixes (and even more are on the way in #213). There you should read [`README.md`](https://github.com/romcal/romcal/blob/_F-TypeScript/README.md) and the files in the [`docs/`](https://github.com/romcal/romcal/tree/_F-TypeScript/docs) folder. Else I could suggest you only to read through the code (it contains many comments) in the [`src/`](https://github.com/romcal/romcal/tree/_F-TypeScript/src) folder.

**Comment by sdegutis** [comment]
I suggest writing down the desired algorithm(s) in high-level pseudocode, so that others can help out more easily by having a clear destination of where to go from the current codebase (any branch). Otherwise, with myself for example, I'm just going to be looking at code I don't understand, and unsure of what to change, and what to change it to.

**Comment by sdegutis** [comment]
> Although it is a dev version, it contains many, many fixes (and even more are on the way in #213).

This gives me a lot of hesitation using this project. You're saying there are two branches, and both have some bugs that the other doesn't, and both have some features that the other doesn't. While it's true that it's a free project, and I'm free to not use it, this instability kind of hurts my confidence in how to use it correctly, or whether that's possible. Just offering this as constructive criticism.

**Comment by tukusejssirs** [comment]
@sdegutis, I am sorry for your negative experience. I am not a coder; I try to tend the issues, localisation and celebration definitions in calendar. I know I can improve some stuff in these areas though.

We have currently one (more or less) active coder—@emagnier. He is currently busy with work / other projects / whatever else, however, he wants to release a new dev version (v2.0.0) of romcal by merging `_F-TypeScript` branch into `dev`, as you may know. In `_F-TypeScript` branch, romcal is much more improved that the current stable version v1.3.0 in `master`.

> I suggest writing down the desired algorithm(s) in high-level pseudocode, so that others can help out more easily by having a clear destination of where to go from the current codebase (any branch). Otherwise, with myself for example, I'm just going to be looking at code I don't understand, and unsure of what to change, and what to change it to.

I agree with you, however, you should open a separate issue (or later even a PR) for this. There we could discuss how we could implement this.

Nevertheless, the `_F-TypeScript` branch have some documentation in the [`docs/`](https://github.com/romcal/romcal/tree/_F-TypeScript/docs) folder, albeit it is not perfect yet. If you haven’t does so yet, you should check it out. There is also the [`README.md`](https://github.com/romcal/romcal/blob/_F-TypeScript/README.md) which also contains some useful information.

> This gives me a lot of hesitation using this project. You're saying there are two branches, and both have some bugs that the other doesn't, and both have some features that the other doesn't. While it's true that it's a free project, and I'm free to not use it, this instability kind of hurts my confidence in how to use it correctly, or whether that's possible. Just offering this as constructive criticism.

Critisism accepted. Could elaborate what features do you miss from the `_F-TypeScript` branch which is implemented in the version from `master`? Same for the bugs you find in the `_F-TypeScript` branch. It might be better to open a separate branch for these.

Thanks in advance! :smiley:

**Comment by emagnier** [comment]
Currently the main active branches are `master` and `_F-TypeScript`. 

- Romcal on the `master` branch is stable. Not so fast but stable (I'm using it in production).
- `_F-TypeScript` is a working branch that now aims to be merged soon in `dev` (and might then open to a new release a bit after). It will be a major release and a big step forward `master`, but I recognise we still didn't make too much effort on runtime performances.

@pejulian and I are the main developers on this project. At some point we decided to migrate the codebase from JavaScript to TypeScript. He did a lot of work first, then I took over. The problem I see is we set the bar too high for the time we have for this project. This was not only a rewrite, but also a refactoring and a redesign of the APIs and data schema that are generated from this library. And a lot of calendars and localisation data have been also refactored and completed.

By "fixes", I guess @tukusejssirs means "improvements". He indeed made a lot of great contributions on improving and consolidating data in calendar definitions (dates, metadata, remove duplicates...) and localisation. I guess we have to differentiate "bugs" or issue from the calendar definitions (wrong date or missing/incorrect metadata for a specific liturgical day), and bugs from the core codebase.

Today, the blockers are mainly the design and implementation of some new features that need to be completed before a release. But I think the new codebase is pretty stable, and even more than before thanks to TypeScript typing and a lot of additional tests (for sure a testing phase will be still necessary). I have in mind to remove some of the new features that are blocking the release. As soon as we release a new version, it will help future contributions (I am aware that currently it's difficult to know where and how to contribute), and this will help to move on next challenges (including runtime performance). I see a window towards the end of next week to make it happen (at least a beta release), I'll try to stick on it.

And yes documentation has also been greatly reworked and completed. For sure there are still missing parts and it's not perfect (I'm not a native English speaker), but I hope it could help a bit people to use romcal. If you have any inputs on that side too, it's also more than welcome. Thanks to share us your inputs and feelings about this project, this is very useful to us 👍

**Comment by emagnier** [comment]
FYI we are working on a new romcal `v3`. 
Indeed we reached a glass ceiling on v2 in terms of performance and the ability to introduce new features. In this v3, the codebase architecture has been redesigned, and 95% of the core codebase has been rewritten. And as a result, romcal v3 is now up to 10x faster (and olso including a few new features). You can have a look at the [changelog](https://github.com/emagnier/romcal/blob/v3/CHANGELOG.md).

As soon as we deliver a first v3-alpha package (as NPM), I'll invite you to test again romcal (I'll drop a note here when it's ready). Thanks for your patience 😉

**Comment by emagnier** [comment]
A new version 3 of romcal as been published on NPM (as alpha version). 

The documentation isn't fully updated yet, but there are at least basic instructions on the main README.md, on how to use this new version of romcal: https://github.com/romcal/romcal/blob/dev/README.md#getting-started

Feel free to test it and provide us feedback. Thank you!
